src/python/pants/backend/python/rules/pytest_coverage.py

```diff
@@ -0,0 +1,414 @@
+    addresses_and_test_results: Tuple[AddressAndTestResult, ...]
+
+
+@dataclass(frozen=True)
+class MergedCoverageData:
+    coverage_data: Digest
+
+
+@rule(name="Merge coverage reports")
+async def merge_coverage_data(
+    data_batch: PytestCoverageDataBatch,
+    transitive_targets: TransitiveHydratedTargets,
+    python_setup: PythonSetup,
+    coverage_setup: CoverageSetup,
+    subprocess_encoding_environment: SubprocessEncodingEnvironment,
+) -> MergedCoverageData:
+    """Takes all python test results and merges their coverage data into a single sql file."""
+    # We start with a bunch of test results, each of which has a coverage data file called `.coverage`
+    # We prefix each of these with their address so that we can write them all into a single pex.
+    coverage_directory_digests = await MultiGet(
+        Get[Digest](
+            DirectoryWithPrefixToAdd(
+                directory_digest=result.test_result.coverage_data.digest,  # type: ignore[attr-defined]
+                prefix=result.address.path_safe_spec,
+            )
+        )
+        for result in data_batch.addresses_and_test_results
+        if result.test_result is not None and result.test_result.coverage_data is not None
+    )
+    sources = await Get[SourceFiles](
+        AllSourceFilesRequest(
+            (ht.adaptor for ht in transitive_targets.closure), strip_source_roots=False
+        )
+    )
+    sources_with_inits_snapshot = await Get[InitInjectedSnapshot](
+        InjectInitRequest(sources.snapshot)
+    )
+    coveragerc = await Get[Coveragerc](
+        CoveragercRequest(HydratedTargets(transitive_targets.closure), test_time=True)
+    )
+    merged_input_files: Digest = await Get(
+        Digest,
+        DirectoriesToMerge(
+            directories=(
+                *coverage_directory_digests,
+                sources_with_inits_snapshot.snapshot.directory_digest,
+                coveragerc.digest,
+                coverage_setup.requirements_pex.directory_digest,
+            )
+        ),
+    )
+
+    prefixes = [
+        f"{result.address.path_safe_spec}/.coverage"
+        for result in data_batch.addresses_and_test_results
+    ]
+    coverage_args = ["combine", *prefixes]
+    request = coverage_setup.requirements_pex.create_execute_request(
+        python_setup=python_setup,
+        subprocess_encoding_environment=subprocess_encoding_environment,
+        pex_path=f"./{coverage_setup.requirements_pex.output_filename}",
+        pex_args=coverage_args,
+        input_files=merged_input_files,
+        output_files=(".coverage",),
+        description=f"Merge coverage reports.",
+    )
+
+    result = await Get[ExecuteProcessResult](ExecuteProcessRequest, request)
+    return MergedCoverageData(coverage_data=result.output_directory_digest)
+
+
+@dataclass(frozen=True)
+class PytestCoverageData(CoverageData):
+    digest: Digest
+
+    @property
+    def batch_cls(self) -> Type["PytestCoverageDataBatch"]:
```

Comment:
Since the line below works this line should too. I think you can drop the str arg now in favor of the raw type since PytestCoverageDataBatch is now defined above this class in the file.